526 start date required for meal plan creation and editing

RLCASTRO opened the pull request:
Describe the technical changes contained in this PR
Creating a new field to hold the meal plan start date to help track different meal plans.

Previous behaviour
No start date was asked when creating a new meal plan

New behaviour
[user/admin/designer] will be able to select a start date for the meal planner

Related issues addressed by this PR
Fixes https://github.com/CivicTechFredericton/mealplanner/issues/526

**Have the following been addressed?**
- [ ] Have test cases been created for all of the changes?
- [ ] Do all of the test cases pass?
- [ ] Has the testing been done using the default docker-compose environment?
- [ ] Are documentation changes required?
- [ ] Does this change break or alter existing behaviour?

RLCASTRO committed:
created the new migration file to add a new startDate field in the meal_plan table

RLCASTRO committed:
Created the new migration file to ALTER the meal_plan table by adding a new field to hold the startDate

RLCASTRO committed:
Installed new dependencies

RLCASTRO committed:
Added new <DatePicker> component and updated the initState, and handleClose function

RLCASTRO committed:
updating the generated files for the new column in the database

RLCASTRO committed:
renaming the migration file

RLCASTRO committed:
Creted a useState to hold the startDate, and added to the initialState and to the handleClose and createMealPlan functions

RLCASTRO committed:
updated the mealPlansQuery to return the startdate value

RLCASTRO committed:
updated the mealPlansQuery to return the startdate value

RLCASTRO committed:
Created a date object with the mealplan.startdate and conditionally rendered a <Chip> component displaying the formated date inside the meal plan card.

RLCASTRO committed:
Updated the GraphQL query to include the startdate and console.log for now

shanthisa commented on backend/migrations/014-526-start-date-required-for-meal-plan.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE app.meal_plan ADD COLUMN startDate TEXT
```
This was part of the earlier PR that was approved and merged. So need to rebase with v2 and not show this as a change in this PR

RLCASTRO replied:
This was solved with the migration files 000012

shanthisa commented on mealplanner-ui/src/pages/MealPlans/MealPlanCard.tsx:
```diff
@@ -73,6 +75,10 @@ export const MealPlanCard = (props: MealPlanCardProps) => {
     };
+
+    useEffect(() => {
+      console.log(mealplan);
```
Please remove console.log

RLCASTRO replied:
done, thank you.

shanthisa commented on mealplanner-ui/src/pages/MealPlans/MealPlanHeader.tsx:
```diff
@@ -61,7 +62,13 @@ export const MealPlanHeader: React.FC<HeaderProps> = ({ mealPlan }) => {
   const navigate = useNavigate();
 
+    useEffect(() => {
+      console.log(data);
```
Please remove console.log

RLCASTRO committed:
updating the field name from startdate to startDate and type from string to date

RLCASTRO committed:
updating the text to be smaller

RLCASTRO committed:
updating the startdate field to startDate in the state.ts, and its generated files

RLCASTRO committed:
updating the startdate field to startDate, and its generated files

RLCASTRO committed:
renaming the migration file

RLCASTRO committed:
modifying the updateMealPlanMutation to handle the startDate field, plus its generated files

RLCASTRO committed:
updating the startDate const type

RLCASTRO committed:
implemented a new field to update the mealplan start date. Improoved the logic to handle the fields state from using onBlur to onMouseEnter.

RLCASTRO committed:
solving conflicts when rebasing with v2

RLCASTRO committed:
hiding the DatePicker component for creating new templates

RLCASTRO committed:
hiding the date component from edit meal plan when is a template

RLCASTRO committed:
adding a conditional statement to the onMouseEnter to avoid displaying the select user dropdown menu

RLCASTRO committed:
Created the new migration files to fix the problem of not creating the startDate field in the database after running the migrations

RLCASTRO committed:
updated the Date object to a Dayjs object to stop displaying wrong date

RLCASTRO committed:
added the new feature to edit the startDate. Repositioned a few components and deleted unused code.

RLCASTRO committed:
Updated the CSS to better look improvments and avoid components overlaping each other

RLCASTRO committed:
upadated the typography to no break lines or overflow other components with bigger text

RLCASTRO committed:
Implemented conditional rendering the the date component in the edit meal plan page

shanthisa commented on backend/db_migrations/000011_meal_plan_start_date.down.sql:
```diff
@@ -0,0 +1,3 @@
+begin;
```
You need to change your migration as there is already 11 in the sequence. It was for my issue #592. Please rebase v2. While rebasing resolve the conflict by accepting the latest v2 changes and deleting yours. You may want to keep another copy temporarily elsewhere, create a new migration file and copy paste it to the new migration file.

RLCASTRO replied:
done, thank you.

shanthisa reviewed (changes requested):
Please change the migration file. You need to change your migration as there is already 11 in the sequence. It was for my issue #592. Please rebase v2. While rebasing resolve the conflict by accepting the latest v2 changes and deleting yours. You may want to keep another copy temporarily elsewhere, create a new migration file and copy paste it to the new migration file.

RLCASTRO committed:
renaming the migration files

RLCASTRO committed:
Removing the console.log

shanthisa commented:
![image](https://github.com/CivicTechFredericton/mealplanner/assets/57197486/7a32ce48-4d73-496b-aefb-6d8662e1e796)

The layout should show the date properly. Also the line should not run through the text 'Edit date'. It should appear with background white space where the word 'Edit date' appears. This can be adjusted with css sx settings.

Also the meal plan is showing as Starting: yyyy-mm-dd format. Instead keep it in the format dd-mmm-yyyy format in both meal plan card as well as Edit meal view. eg: 02-Apr-2024.

RLCASTRO committed:
Updated the css so the date component does not shrink in smaller window view sizes

RLCASTRO committed:
Added a date format in the DatePicker component to display date properly, but kept the date format to store the date on the database as numbers only to facilitate retrieving data from the database

RLCASTRO committed:
Updated the date format to display date as DD-MMM-YYYY

RLCASTRO committed:
Date were displaying wrong values. Fixed by updating the date format that goes to the database

RLCASTRO committed:
Updated the date format, and apply some css rules

RLCASTRO committed:
Removing unused code

RLCASTRO committed:
deleting the DatePicker's label

RLCASTRO committed:
Updating the css rules to fix the flexbox structure and fix the component's sizes

RLCASTRO committed:
adjusted the name component to wrap lines with big names

RLCASTRO committed:
adjusted the DatePicker component to not bi smaller then 11rem to avoid overflowing the date

RLCASTRO committed:
Removing unecessary code

RLCASTRO commented:
> ![image](https://private-user-images.githubusercontent.com/57197486/318959655-7a32ce48-4d73-496b-aefb-6d8662e1e796.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMjA2NTksIm5iZiI6MTcxMjMyMDM1OSwicGF0aCI6Ii81NzE5NzQ4Ni8zMTg5NTk2NTUtN2EzMmNlNDgtNGQ3My00OTZiLWFlZmItNmQ4NjYyZTFlNzk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA1VDEyMzIzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNjM5ZGNkOGFmNWMwMjBiYmQyYmE0ZDY0Y2U2NGE1NmUwMzhkNGY1MDI4ZTA4OGI2NjI4YTRhMjA0YjBkMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kamXil1vLzYse7UptwWkW-FRyIu9typlVXmmtXz02ow)
> 
> The layout should show the date properly. Also the line should not run through the text 'Edit date'. It should appear with background white space where the word 'Edit date' appears. This can be adjusted with css sx settings.
> 
> Also the meal plan is showing as Starting: yyyy-mm-dd format. Instead keep it in the format dd-mmm-yyyy format in both meal plan card as well as Edit meal view. eg: 02-Apr-2024.

![image](https://github.com/CivicTechFredericton/mealplanner/assets/37277959/416f5e4a-752f-487e-b4e5-7fbf6ae31e8d)
![image](https://github.com/CivicTechFredericton/mealplanner/assets/37277959/46093a23-cafd-48b1-99b3-9bd006ad3d9d)

- Fixed the DatePicker component size
- Fixed the DatePicker placeholder text to DD-MM-YYYY when startDate is empty
- Removed the DatePicker's label
- Updated the meal plan name component to wrap big names into multiple lines
- Fixed other CSS problems

shanthisa commented:
I like the way Edit View has come out well. You had done the UI changes requested. It looks good.
Can you please share the image of Meal Card with the Start Date?
Also, I would like to request a small change. It would be good to label it as 'Start Date: dd-mmm-yyyy" instead of "Starting Date:.."

shanthisa commented on mealplanner-ui/src/pages/MealPlans/CreateMealPlan.tsx:
```diff
@@ -155,7 +162,23 @@ export const CreateMealPlan = ({ connection, refetch }: { connection: string, re
+              {planType === "mealPlan" && (
+                <LocalizationProvider dateAdapter={AdapterDayjs}>
+                  <DatePicker
+                    label="Starting Date"
```
Instead of 'Starting Date'. It will be nice to have it as 'Start Date:'

RLCASTRO replied:
Now it's done. Thank you.

shanthisa commented on mealplanner-ui/src/pages/MealPlans/MealPlanCard.tsx:
```diff
@@ -163,6 +166,11 @@ export const MealPlanCard = (props: MealPlanCardProps) => {
           />
+          {startDate && (
+            <div style={{ display: "flex", justifyContent: "center", alignItems: "center" }}>
+              <Chip label={"Starting: " + startDate.format("DD-MMM-YYYY")} />
```
In hindsight, since 'Starting on' is correct grammatically and 'Starting' sounds incomplete, I thought we could just have it as 'Start Date'. Moreover, don't you think that having it as a tooltip on hover would make it look slicker? That is just the date alone to be shown on the Chip, and on hover, the text 'Start Date' is shown as a tooltip.

RLCASTRO replied:
That's a great idea. thank you.

shanthisa reviewed (changes requested):
Overall looks great. Just a small cosmetic change requested.

RLCASTRO committed:
Updated the meal planner card to display the start date itself and a tooltip that says it is a the start date on hover

RLCASTRO commented:
Updated the meal plan card to display the date only, and on hover, it displays a tooltip with "Start Date".

- Updated Meal Plan Card

![image](https://github.com/CivicTechFredericton/mealplanner/assets/37277959/12b5c4b5-5a06-47f7-8e91-3fe88e9ca2f0)

RLCASTRO committed:
Updated the label to match all the other labels and keep in a standard

shanthisa reviewed (approved):
LGTM 👍
